### PR TITLE
Add an end to end test to show the entire system working

### PR DESCRIPTION
**Pull request by ghost**
# Context

#71 added a system of factories to help us reflect the client/server split in our testing. This is the first PR to use that system, and it uses it to implement our first proper end to end test.

## Changes proposed in this pull request

* Adds an end to end test
    * Uses a new `e2e` schema for now to avoid conflict with legacy tests
* Adds a `docker` pytest mark to indicate tests that require resources in Docker
    * Updates the just command to allow `just test local`, `just test docker` and `just test`
* Adds a `query_to_cluster_entities()` function to create `ClusterEntities` for `query()`s from the real Matchbox to diff nicely with the factory system
* Fixed minor bug in the diff so that successful matches contribute to the overall mean match ratio
* Minor changes to client functions:
    * `select()` can suppress non-index warning
    * `trim()` added to cleaning steps

## Guidance to review

Confirm you're happy with the interim `e2e` schema solution while other tests migrate.

Other that reading and confirming the e2e test passes, consider bringing down all the Docker resources and running `just test local`.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] New and existing unit tests pass locally with my changes
- [x] I've changed or updated relevant documentation

### CODE REVIEW

File: src/matchbox/client/helpers/selector.py [hunk 1]
```diff
@@ -24,14 +24,20 @@ class Selector(BaseModel):
...
     """From one engine, builds and verifies a list of selectors.
 
     Args:
-        selection: full source names and optionally a subset of columns to select
-        engine: the engine to connect to the data warehouse hosting the source.
+        selection: Full source names and optionally a subset of columns to select
+        engine: The engine to connect to the data warehouse hosting the source.
             If not provided, will use a connection string from the
             `MB__CLIENT__DEFAULT_WAREHOUSE` environment variable.
+        non_indexed: Whether you intend to select non-indexed columns. Will raise a
```
Comment (leo-mazzone) [inline src/matchbox/client/helpers/selector.py:36]: should it not... raise a warning if False? I think it would be more intuitive if this were a `only_indexed=True` argument

File: test/fixtures/db.py [hunk 2]
```diff
@@ -386,12 +388,29 @@ def s3(aws_credentials: None) -> Generator[S3Client, None, None]:
...
-# Mock API
+# API, mocked and Docker
 
 
 @pytest.fixture(scope="function")
 def matchbox_api() -> Generator[MockRouter, None, None]:
+    """Client for the mocked Matchbox API."""
     with respx.mock(
-        base_url=client_settings.api_root, assert_all_called=True
+        base_url="http://localhost:8000", assert_all_called=True
```
Comment (leo-mazzone) [inline test/fixtures/db.py:398]: why hardcode this? Doesn't seem like a good idea. It also makes it more difficult to redirect this to port 8080 when one wants to debug the API

File: test/common/factories/test_entity_factory.py [hunk 3]
```diff
@@ -436,8 +470,7 @@ def test_source_to_results_conversion():
     assert results1.similarity_ratio(results3) == 0.0
 
     # Test error case for missing dataset
```
Comment (leo-mazzone) [inline test/common/factories/test_entity_factory.py:472]: not a test error case

File: test/common/factories/test_entity_factory.py [hunk 4]
```diff
@@ -346,6 +346,40 @@ def test_probabilities_to_results_entities(
...
+                make_cluster_entity(2, "d1", ["3", "4"]),  # Will be partially matched
+            ],
+            [
+                make_cluster_entity(1, "d1", ["1", "2"]),  # Same as in expected
+                make_cluster_entity(3, "d1", ["3", "5"]),  # Partial match with entity 2
+            ],
+            True,
+            False,
+            {
+                "mean_similarity": pytest.approx(5 / 9, rel=1e-2),
```
Comment (leo-mazzone) [inline test/common/factories/test_entity_factory.py:364]: why 5 / 9?

File: src/matchbox/common/factories/entities.py [hunk 5]
```diff
@@ -449,19 +450,62 @@ def to_cluster_entity(self, *names: str) -> ClusterEntity:
...
+        query: A PyArrow table or DataFrame representing a query result
+        source_pks: Mapping of source names to primary key column names
+
+    Returns:
+        A set of ClusterEntity objects
+    """
+    if isinstance(query, pa.Table):
+        query = query.to_pandas()
+
+    must_have_columns = set(["id"] + list(source_pks.values()))
```
Comment (leo-mazzone) [inline src/matchbox/common/factories/entities.py:486]: multiple pks?

Comment (leo-mazzone) [reply src/matchbox/common/factories/entities.py:486]: ah, sorry, one per source right?

File: test/e2e/test_e2e.py [hunk 6]
```diff
@@ -0,0 +1,533 @@
...
+                full_name="e2e.cdms",
+                engine=warehouse_engine,
+                features=(
+                    features["crn"],
+                    features["cdms"],
+                ),
+                n_true_entities=n_true_entities,
+                repetition=1,  # Duplicate all rows
+            ),
+        )
```
Comment (leo-mazzone) [inline test/e2e/test_e2e.py:104]: I wonder whether this is a good time to start divorcing this from cdms and duns names? The former is an obscure legacy system no one remembers anymore, and the latter is a commercial product

File: test/e2e/test_e2e.py [hunk 7]
```diff
@@ -0,0 +1,533 @@
...
+
+        # Use a separate schema to avoid conflict with legacy test data
+        # TODO: Remove once legacy tests are refactored
+        with warehouse_engine.connect() as conn:
+            conn.execute(text("create schema e2e;"))
+            conn.commit()
+
+        # Setup code - Create tables in warehouse
+        for source_name, source_testkit in self.linked_testkit.sources.items():
```
Comment (leo-mazzone) [inline test/e2e/test_e2e.py:119]: Will this be part of the later refactor? This is very much postgres-specific

File: test/e2e/test_e2e.py [hunk 8]
```diff
@@ -0,0 +1,533 @@
...
+
+            return df
+
+        # === DEDUPLICATION PHASE ===
+        deduper_names = {}
+
+        for source_name, source_testkit in self.linked_testkit.sources.items():
+            # Get prefix for column names based on source
+            schema, table = source_name.replace('"', "").split(".")
+            prefix = f"{schema}_{table}_"
```
Comment (leo-mazzone) [inline test/e2e/test_e2e.py:206]: could you instead replace dots with underscores and add an underscore at the end? Not sure I've thought of all the edge cases, but would make this more general than a Postgres warehouse

File: test/e2e/test_e2e.py [hunk 9]
```diff
@@ -0,0 +1,533 @@
...
+
+            # Query data from the source
+            # PK included then dropped to create ClusterEntity objects for later diff
+            source_select = select(
+                {
+                    source_name: ["pk"]
+                    + [col.alias for col in source_testkit.source.columns]
```
Comment (leo-mazzone) [inline test/e2e/test_e2e.py:213]: that's now what the alias is for I believe, you should use the name instead

File: test/e2e/test_e2e.py [hunk 10]
```diff
@@ -0,0 +1,533 @@
...
+            select(
+                {duns_source: ["pk", "company_name", "duns"]},
+                engine=self.warehouse_engine,
+                non_indexed=True,
+            ),
+            select(
+                {cdms_source: ["pk", "crn", "cdms"]},
+                engine=self.warehouse_engine,
+                non_indexed=True,
+            ),
```
Comment (leo-mazzone) [inline test/e2e/test_e2e.py:498]: these are using the same engine, can all go in the same select statement

**Review by leo-mazzone — APPROVED** [review]
It's good to reap the benefits of your foundational work on testkits. This reads great, well  done. Incidentally, I could run all local tests (245) in 11 seconds.
